src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,257 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            if (argType.IsErrorType())
+            {
+                return false;
+            }
+
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            var lookupResult = LookupResult.GetInstance();
+            if (PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out indexerAccess, argType, diagnostics, ref useSiteInfo))
+            {
+                _ = GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, diagnostics, syntax: syntax);
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            if (TryFindIndexOrRangeIndexerPattern(syntax, lookupResult, receiverOpt: null, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo))
+            {
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            lookupResult.Free();
+            return false;
+        }
+
+        private bool PerformPatternIndexerLookup(
+            SyntaxNode syntax, TypeSymbol receiverType, LookupResult lookupResult,
+            [NotNullWhen(true)] out BoundIndexerAccess? result, TypeSymbol indexerArgumentType, BindingDiagnosticBag diagnostics, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            result = null;
+
+            LookupMembersInType(
+                lookupResult,
+                receiverType,
+                WellKnownMemberNames.Indexer,
+                arity: 0,
+                basesBeingResolved: null,
+                LookupOptions.Default,
+                originalBinder: this,
+                diagnose: false,
+                ref useSiteInfo);
+
+            if (lookupResult.IsMultiViable)
+            {
+                var indexerGroup = ArrayBuilder<PropertySymbol>.GetInstance();
+                foreach (Symbol symbol in lookupResult.Symbols)
+                {
+                    Debug.Assert(symbol.IsIndexer());
+                    indexerGroup.Add((PropertySymbol)symbol);
+                }
+
+                var analyzedArguments = AnalyzedArguments.GetInstance();
+                analyzedArguments.Arguments.Add(new BoundIndexOrRangeIndexerPatternValuePlaceholder(syntax, indexerArgumentType));
+                BoundExpression receiver = new BoundImplicitReceiver(syntax, receiverType);
+                var bindingDiagnostics = BindingDiagnosticBag.Create(diagnostics);
+                var boundAccess = BindIndexerOrIndexedPropertyAccess(syntax, receiver, indexerGroup, analyzedArguments, bindingDiagnostics);
+                if (boundAccess is BoundIndexerAccess { ResultKind: LookupResultKind.Viable } indexerAccess &&
+                    indexerAccess.Indexer.GetMethod is { } getMethod && IsAccessible(getMethod, ref useSiteInfo))
```

Comment:
>IsAccessible(getMethod, ref useSiteInfo) [](http://example.com/codeflow?start=72&length=40)

What are we going to do if it's not accessible? If there's no fallback, it somewhat feels like we should let it bind then error about the method not being visible.

Reply:
>What are we going to do if it's not accessible? 

Then we'll look for the implicit support. We don't want any diagnostics here except for when we successfully found a viable member. Beyond that I think this goes into "diagnostics clarity" bucket which I think can be done later.